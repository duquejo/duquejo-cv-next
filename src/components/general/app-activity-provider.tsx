'use client';

import { ReactNode, useState } from 'react';
import {
  Sheet,
  SheetContent,
  SheetDescription,
  SheetFooter,
  SheetHeader,
  SheetTitle,
} from '@/components/ui/sheet';
import Link from 'next/link';
import { EventSkeleton } from '@/components/events/event-skeleton';
import useSWR from 'swr';
import { getEvents } from '@/actions/github';
import { Event } from '@/interfaces';
import { EventCard } from '@/components/events/event-card';
import { useTranslations } from 'next-intl';

interface Props {
  children: ReactNode;
  defaultOpen?: boolean;
}

export const AppActivityProvider = ({ children, defaultOpen = false }: Props) => {
  const t = useTranslations('Sidebar.complementary.activity.dialog');
  const [loadEvents, setLoadEvents] = useState(false);

  const { data, isLoading } = useSWR<Event[]>(loadEvents ? 'events' : null, getEvents);

  const onToggleChange = async (open: boolean) => {
    if (open) {
      return setLoadEvents(true);
    }
    setLoadEvents(false);
  };

  return (
    <Sheet key="activity" onOpenChange={onToggleChange} defaultOpen={defaultOpen}>
      {children}
      <SheetContent role="dialog" className="flex flex-col justify-between w-10/12">
        <SheetHeader>
          <SheetTitle>{t('title')}</SheetTitle>
          <SheetDescription>{t('description')}</SheetDescription>
        </SheetHeader>
        <div className="flex-1 flex flex-col gap-2 overflow-y-auto">
          {isLoading ? (
            <EventSkeleton rounds={2} />
          ) : (
            <div role="list" className="flex flex-col gap-2 overflow-y-auto list-none">
              {!data || data.length === 0 ? (
                <p data-testid="empty" className="text-muted-foreground text-sm">
                  {t('empty')}
                </p>
              ) : (
                data.map((event) => <EventCard key={event.id} {...event} />)
              )}
            </div>
          )}
        </div>
        <SheetFooter>
          <Link
            href="https://github.com/duquejo"
            target="_blank"
            className="text-xs text-muted-foreground"
          >
            {t('button')}
          </Link>
        </SheetFooter>
      </SheetContent>
    </Sheet>
  );
};
