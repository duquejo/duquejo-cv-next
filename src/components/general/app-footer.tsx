'use client';

import Link from 'next/link';
import { Social } from '@/interfaces';
import { Github, Linkedin } from 'lucide-react';

export const SOCIAL_DATA: Social[] = [
  {
    name: 'GitHub',
    link: 'https://github.com/duquejo',
    icon: <Github />,
  },
  {
    name: 'LinkedIn',
    link: 'https://www.linkedin.com/in/duquejo/',
    icon: <Linkedin />,
  },
];

export const AppFooter = () => {
  return (
    <footer className="border-t flex p-3 items-center justify-center gap-x-8">
      {SOCIAL_DATA.map((social) => (
        <Link
          href={social.link}
          key={social.name}
          target="_blank"
          title={social.name}
          aria-label={social.name}
        >
          {social.icon}
        </Link>
      ))}
      <span className="font-semibold text-sm">© {new Date().getFullYear()} José Duque</span>
    </footer>
  );
};
