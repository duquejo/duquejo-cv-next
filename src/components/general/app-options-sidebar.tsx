import { ThemeToggle } from '@/components/general/theme-toggle';
import { SidebarTrigger } from '@/components/ui/sidebar';

export const AppOptionsSidebar = () => {
  return (
    <nav className="lg:w-14 lg:max-w-14 lg:h-screen lg:relative w-auto fixed right-0 lg:mr-1">
      <div className="flex flex-row gap-2 bg-sidebar rounded-lg p-2 m-2 lg:mx-0 lg:mb-2 lg:flex-col lg:fixed lg:bottom-0 lg:border lg:p-1">
        <SidebarTrigger
          variant="outline"
          className="h-8 w-8 lg:h-10 lg:w-10"
          title="Toggle sidebar"
        />
        <ThemeToggle
          variant="outline"
          className="h-8 w-8 lg:h-10 lg:w-10"
          title="Change color theme"
        />
      </div>
    </nav>
  );
};
