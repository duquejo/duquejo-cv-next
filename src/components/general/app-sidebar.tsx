import {
  Sidebar,
  SidebarContent,
  SidebarGroup,
  SidebarGroupContent,
  SidebarGroupLabel,
  SidebarHeader,
  SidebarMenu,
  SidebarSeparator,
} from '@/components/ui/sidebar';
import { MenuItem } from '@/interfaces';
import { Calendar, Gamepad, Inbox, Music, Pencil, Projector, Send } from 'lucide-react';
import { MenuAvatar } from '@/components/resume/menu-avatar/MenuAvatar';
import { MenuItems } from '@/components/resume/menu-items/MenuItems';

// Menu items.
export const professionalItems: MenuItem[] = [
  {
    title: 'Resume√©',
    url: '/',
    icon: <Inbox />,
  },
  {
    title: 'Services',
    url: '/career/services',
    icon: <Calendar />,
  },
  {
    title: 'Latest projects',
    url: '/career/projects',
    icon: <Projector />,
  },
];

export const lastItems: MenuItem[] = [
  {
    title: 'My blog',
    url: '/blog',
    icon: <Pencil />,
  },
  {
    title: 'Contact',
    url: '/contact',
    icon: <Send />,
  },
];

export const hobbiesItems: MenuItem[] = [
  {
    title: 'Music production',
    url: '/hobbies/music-production',
    icon: <Music />,
  },
  {
    title: 'Game development',
    url: '/hobbies/game-development',
    icon: <Gamepad />,
  },
];

export const AppSidebar = () => {
  return (
    <Sidebar collapsible="icon" variant="floating" role="navigation" tabIndex={0}>
      <SidebarHeader>
        <div className="flex flex-col place-items-center gap-y-2">
          <MenuAvatar />
        </div>
      </SidebarHeader>
      <SidebarContent className="justify-between">
        <div>
          <SidebarGroup>
            <SidebarGroupLabel>Professional career</SidebarGroupLabel>
            <SidebarGroupContent>
              <SidebarMenu>
                <MenuItems items={professionalItems} />
              </SidebarMenu>
            </SidebarGroupContent>
          </SidebarGroup>
          <SidebarGroup>
            <SidebarGroupLabel>Hobbies</SidebarGroupLabel>
            <SidebarGroupContent>
              <SidebarMenu>
                <MenuItems items={hobbiesItems} />
              </SidebarMenu>
            </SidebarGroupContent>
          </SidebarGroup>
        </div>
        <div>
          <SidebarSeparator />
          <SidebarGroup>
            <SidebarGroupContent>
              <SidebarMenu>
                <MenuItems size="sm" items={lastItems} />
              </SidebarMenu>
            </SidebarGroupContent>
          </SidebarGroup>
        </div>
      </SidebarContent>
    </Sidebar>
  );
};
