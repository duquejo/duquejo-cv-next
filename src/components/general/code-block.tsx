import { Light as SyntaxHighlighter } from 'react-syntax-highlighter';
import ts from 'react-syntax-highlighter/dist/esm/languages/hljs/typescript';
import atomDark from 'react-syntax-highlighter/dist/cjs/styles/hljs/atom-one-dark-reasonable';

SyntaxHighlighter.registerLanguage('javascript', ts);

interface CodeBlockProps {
  code: string;
}

export const CodeBlock = ({ code }: CodeBlockProps) => {
  return (
    <SyntaxHighlighter
      style={atomDark}
      showLineNumbers
      customStyle={{ margin: 0, borderRadius: 0, fontSize: 14 }}
    >
      {code}
    </SyntaxHighlighter>
  );
};
