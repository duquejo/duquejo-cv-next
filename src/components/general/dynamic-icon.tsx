import type { ComponentType } from 'react';
import dynamic from 'next/dynamic';
import { LoaderCircle } from 'lucide-react';

interface Props {
  iconName: string;
  [key: string]: any; // eslint-disable-line @typescript-eslint/no-explicit-any
}
type LucideIconsModule = Record<string, ComponentType<unknown>>;

export const DynamicIcon = ({ iconName, ...props }: Props) => {
  const Icon = dynamic(
    () =>
      import('lucide-react').then((module) => {
        const iconsModule = module as unknown as LucideIconsModule;
        const component = iconsModule[iconName];
        if (!component) {
          throw new Error(`Icon ${iconName} not found`);
        }
        return component;
      }),
    {
      loading: () => (
        <LoaderCircle size="70" strokeWidth={1} className="text-secondary-foreground" />
      ),
    },
  );

  return <Icon {...props} />;
};
