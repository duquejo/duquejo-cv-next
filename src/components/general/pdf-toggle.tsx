'use client';

import { Button, type ButtonProps } from '@/components/ui/button';
import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/popover';
import { ImageDown } from 'lucide-react';
import { cn } from '@/lib';
import { PdfForm } from '@/components/pdf/pdf-form';
import { useState } from 'react';

interface Props {
  className?: string;
  buttonTitle: string;
  formDescription: string;
  formTitle: string;
  formButtonText: string;
  formButtonTextLoading: string;
  variant?: ButtonProps['variant'];
}

export const PdfGeneratorToggle = ({
  formDescription,
  formButtonText,
  formTitle,
  buttonTitle,
  formButtonTextLoading,
  className,
  variant,
}: Props) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);

  return (
    <Popover key="pdf" open={isPopoverOpen}>
      <PopoverTrigger asChild>
        <Button
          variant={variant}
          size="icon"
          className={cn('h-10 w-10', className)}
          onClick={() => setIsPopoverOpen(true)}
        >
          <ImageDown className="h-8 w-8 lg:h-10 lg:w-10" aria-label={buttonTitle} />
          <span className="sr-only">{buttonTitle}</span>
        </Button>
      </PopoverTrigger>
      <PopoverContent
        role="dialog"
        aria-label={formTitle}
        onInteractOutside={() => setIsPopoverOpen(false)}
        onEscapeKeyDown={() => setIsPopoverOpen(false)}
      >
        <h2 id="form-title" className="font-medium leading-none mb-2 text-sm">
          {formTitle}
        </h2>
        <p id="form-description" className="text-muted-foreground text-xs">
          {formDescription}
        </p>
        <PdfForm
          onSubmitFinish={() => setIsPopoverOpen(false)}
          button={formButtonText}
          buttonLoading={formButtonTextLoading}
        />
      </PopoverContent>
    </Popover>
  );
};
