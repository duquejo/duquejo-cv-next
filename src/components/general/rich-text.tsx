// @flow
import * as React from 'react';
import { ReactNode } from 'react';

type Tag = 'b' | 'br';

interface Props {
  children(tags: Record<Tag, (chunks: ReactNode) => ReactNode>): ReactNode;
}
export const RichText = ({ children }: Props) => {
  return children({
    b: (chunks: ReactNode) => <b>{chunks}</b>,
    br: () => <br />,
  });
};
