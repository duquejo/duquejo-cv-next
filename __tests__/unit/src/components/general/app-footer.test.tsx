import { render, type RenderResult, screen } from '@testing-library/react';
import { AppFooter, SOCIAL_DATA as social } from '@/components/general/app-footer';

describe('<AppFooter /> tests', () => {
  let component: RenderResult;

  beforeEach(() => {
    component = render(<AppFooter />);

    const copyDate = component.container.querySelector('span');
    if (copyDate) {
      copyDate.textContent = '© 1992 José Duque';
    }
  });

  it('Should match the snapshot', () => {
    // Arrange & Act
    const footer = screen.getByRole('contentinfo');

    // Assert
    expect(component.container).toMatchSnapshot();
    expect(footer).toBeInTheDocument();
    expect(footer).toHaveTextContent('© 1992 José Duque');

    social.map((s) => {
      const link = screen.getByRole('link', { name: s.name });

      expect(link).toBeInTheDocument();
      expect(link).toHaveAttribute('href', s.link);
      expect(link).toHaveAttribute('title', s.name);
      expect(link).toHaveAttribute('aria-label', s.name);
      expect(link).toHaveAttribute('target', '_blank');
      expect(link).toContainHTML('svg');
    });
  });
});
