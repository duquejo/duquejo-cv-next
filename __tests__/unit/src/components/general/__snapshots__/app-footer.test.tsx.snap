// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Footer /> tests > Should match the snapshot 1`] = `
<div>
  <footer
    class="flex items-center justify-center gap-x-5 py-3 fixed bottom-0 w-full bg-sidebar sm:bg-transparent sm:relative"
  >
    <div
      class="flex items-center justify-center gap-x-5"
    >
      <a
        aria-label="GitHub"
        class="focus-visible:ring-1 hover:text-primary"
        href="https://github.com/duquejo"
        target="_blank"
        title="GitHub"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-github"
          fill="none"
          height="23"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="23"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
          />
          <path
            d="M9 18c-4.51 2-5-2-7-2"
          />
        </svg>
      </a>
      <a
        aria-label="LinkedIn"
        class="focus-visible:ring-1 hover:text-primary"
        href="https://www.linkedin.com/in/duquejo/"
        target="_blank"
        title="LinkedIn"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-linkedin"
          fill="none"
          height="23"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="23"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 8a6 6 0 0 1 6 6v7h-4v-7a2 2 0 0 0-2-2 2 2 0 0 0-2 2v7h-4v-7a6 6 0 0 1 6-6z"
          />
          <rect
            height="12"
            width="4"
            x="2"
            y="9"
          />
          <circle
            cx="4"
            cy="4"
            r="2"
          />
        </svg>
      </a>
      <a
        aria-label="E-mail"
        class="focus-visible:ring-1 hover:text-primary"
        href="mailto:contact@duquejo.com"
        target="_blank"
        title="E-mail"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-mail"
          fill="none"
          height="23"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="23"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m22 7-8.991 5.727a2 2 0 0 1-2.009 0L2 7"
          />
          <rect
            height="16"
            rx="2"
            width="20"
            x="2"
            y="4"
          />
        </svg>
      </a>
    </div>
    <span
      class="border-l pl-5 font-semibold text-xs"
    >
      © 1992 José Duque
    </span>
  </footer>
</div>
`;
